crates/bifrost/src/providers/replicated_loglet/network.rs

```diff
@@ -195,34 +221,75 @@ impl RequestPump {
     async fn get_loglet<T: TransportConnect>(
         &self,
         provider: &ReplicatedLogletProvider<T>,
+        peer_version: &PeerMetadataVersion,
         header: &CommonRequestHeader,
     ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
-        let loglet = match provider.get_active_loglet(header.log_id, header.segment_index) {
-            Some(loglet) if loglet.params().loglet_id == header.loglet_id => loglet,
-            Some(_) => return Err(SequencerStatus::LogletIdMismatch),
-            None => {
-                let logs = self.metadata.logs();
-                let chain = logs
-                    .chain(&header.log_id)
-                    .ok_or(SequencerStatus::UnknownLogId)?;
-
-                let segment = chain
-                    .iter()
-                    .rev()
-                    .find(|segment| segment.index() == header.segment_index)
-                    .ok_or(SequencerStatus::UnknownSegmentIndex)?;
-
-                provider
-                    .get_or_create_loglet(
-                        header.log_id,
-                        header.segment_index,
-                        &segment.config.params,
+        let mut current_logs_version = provider.networking().metadata().logs_version();
+        let request_logs_version = peer_version.logs.unwrap_or(current_logs_version);
+
+        loop {
+            if let Some(loglet) = provider.get_active_loglet(header.log_id, header.segment_index) {
+                if loglet.params().loglet_id == header.loglet_id {
+                    return Ok(loglet);
+                }
+
+                return Err(SequencerStatus::LogletIdMismatch);
+            }
+
+            match self.create_loglet(provider, header).await {
+                Ok(loglet) => return Ok(loglet),
+                Err(SequencerStatus::UnknownLogId | SequencerStatus::UnknownSegmentIndex) => {
+                    // possible outdated metadata
+                }
+                Err(status) => return Err(status),
+            }
+
+            if request_logs_version > current_logs_version {
+                match provider
+                    .networking()
+                    .metadata()
+                    .sync(
+                        MetadataKind::Logs,
+                        TargetVersion::Version(request_logs_version),
                     )
-                    .map_err(SequencerStatus::from)?
+                    .await
+                {
+                    Ok(_) => continue,
+                    Err(SyncError::Shutdown(_)) => return Err(SequencerStatus::Shutdown),
+                    Err(SyncError::MetadataStore(err)) => {
+                        tracing::trace!(error=%err, target_version=%request_logs_version, "Failed to sync metadata");
+                        //todo(azmy): make sleep configurable
+                        tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
I've a counter proposal to this. What about we wait for this version to be available instead? (other sync mechanisms will fulfill this when it's possible).

For instance:
`metadata.wait_for_version(MetadataKind::Logs, request_logs_version).await`

If you like, you can also surround this wait with an upper-bound timeout to fail the append if this update never arrived.